File: src/bin/3387.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn max_amount(
        initial_currency: String,
        pairs1: Vec<Vec<String>>,
        rates1: Vec<f64>,
        pairs2: Vec<Vec<String>>,
        rates2: Vec<f64>,
    ) -> f64 {
        fn bellman_ford(d: &mut HashMap<String, f64>, pairs: Vec<Vec<String>>, rates: Vec<f64>) {
            for _ in 0..pairs.len() {
                for (pair, &rate) in pairs.iter().zip(rates.iter()) {
                    let (start, target) = (&pair[0], &pair[1]);
                    let start_val = *d.entry(start.clone()).or_insert(0.0);
                    let target_val = *d.entry(target.clone()).or_insert(0.0);
                    d.insert(start.clone(), start_val.max(target_val / rate));
                    d.insert(target.clone(), target_val.max(start_val * rate));
                }
            }
        }
        let mut d: HashMap<String, f64> = HashMap::from([(initial_currency.clone(), 1.0)]);
        bellman_ford(&mut d, pairs1, rates1);
        bellman_ford(&mut d, pairs2, rates2);
        d[&initial_currency]
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_example1() {
        let initial_currency = "EUR".to_string();
        let pairs1 = vec![
            vec!["EUR".to_string(), "USD".to_string()],
            vec!["USD".to_string(), "JPY".to_string()],
        ];
        let rates1 = vec![2.0, 3.0];
        let pairs2 = vec![
            vec!["JPY".to_string(), "USD".to_string()],
            vec!["USD".to_string(), "CHF".to_string()],
            vec!["CHF".to_string(), "EUR".to_string()],
        ];
        let rates2 = vec![4.0, 5.0, 6.0];
        let result = Solution::max_amount(initial_currency, pairs1, rates1, pairs2, rates2);
        assert_eq!(result, 720.00000);
    }

    #[test]
    fn test_example2() {
        let initial_currency = "ACB".to_string();
        let pairs1 = vec![
            vec!["SFE".to_string(), "Z".to_string()],
            vec!["ACB".to_string(), "SFE".to_string()],
        ];
        let rates1 = vec![9.1, 2.4];
        let pairs2 = vec![
            vec!["SFE".to_string(), "Z".to_string()],
            vec!["ACB".to_string(), "SFE".to_string()],
        ];
        let rates2 = vec![2.8, 5.0];
        let result = Solution::max_amount(initial_currency, pairs1, rates1, pairs2, rates2);
        assert_eq!(result, 1.56000);
    }
}

fn main() {}

File: src/3387/3387.js
/**
 * @param {string} initialCurrency
 * @param {string[][]} pairs1
 * @param {number[]} rates1
 * @param {string[][]} pairs2
 * @param {number[]} rates2
 * @return {number}
 */
var maxAmount = function (initialCurrency, pairs1, rates1, pairs2, rates2) {
  // Функция для выполнения Беллмана-Форда
    const bellmanFord = (amounts, pairs, rates) => {
        const n = pairs.length;

        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                
                const start = pairs[j][0];
                const target = pairs[j][1];
                const rate = rates[j];
                
                const startVal = amounts.get(start) || 0;
                const targetVal = amounts.get(target) || 0;
                
                if (startVal > 0) {
                    amounts.set(target, Math.max(targetVal, startVal * rate));
                }

                if (targetVal > 0) {
                    amounts.set(start, Math.max(startVal, targetVal / rate));
                }
        }
    }
  };

  // Инициализируем суммы
  const amounts = new Map();
  amounts.set(initialCurrency, 1.0);
  bellmanFord(amounts, pairs1, rates1);
  
  const amountsDay1 = new Map(amounts);
  amounts.clear();
  
  amounts.set(initialCurrency, 1.0);
  const pairs2Reverse = pairs2.map(([start, target]) => [target, start]);
  bellmanFord(amounts, pairs2Reverse, rates2);

  let maxAmount = 1.0;
  
  for (const [currency, amount1] of amountsDay1) {
    
    const amount2 = amounts.get(currency) || 0;
    
    if (amount1 > 0 && amount2 > 0) {
        maxAmount = Math.max(maxAmount, amount1 * amount2);
    }
  }

  return Math.round(maxAmount * 100000) / 100000;
};

module.exports = maxAmount;